### PR TITLE
Add spearman correlation

### PR DESCRIPTION
**Pull request by JackKuo666**
Add spearman correlation, Spearman principle comes from [Spearman's rank correlation coefficient](https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient)

### CODE REVIEW

**Comment by schmmd** [comment]
@DeNeutoy would you take a look?

File: allennlp/training/metrics/spearman_correlation.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            self.total_gold_labels = torch.cat((self.total_gold_labels, gold_labels), 0)
+
+    @overrides
+    def get_metric(self, reset: bool = False):
+        """
+        Returns
+        -------
+        The accumulated sample Spearman correlation.
+        """
+        if reset:
```
Comment (DeNeutoy) [inline allennlp/training/metrics/spearman_correlation.py]: This call to `reset` needs to be after you call the metric, otherwise it will be all zeros :)

Comment (DeNeutoy) [reply allennlp/training/metrics/spearman_correlation.py]: Please add a test for this as well

File: allennlp/tests/training/metrics/spearman_correlation_test.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        predictions2 = np.random.randn(1).repeat(num_labels).astype("float32")
+        predictions2 = predictions2[np.newaxis, :].repeat(batch_size, axis=0)
+        labels2 = np.random.randn(1).repeat(num_labels).astype("float32")
+        labels2 = 0.5 * predictions2 + labels2[np.newaxis, :].repeat(batch_size, axis=0)
+
+        # in top 100 batch, 10 number in each batch is different; The last 100 batch, 10 number in each batch is same.
+        predictions_labels_ = [(predictions1, labels1), (predictions2, labels2)]
+
+        stride = 10
```
[the remark this answers is not in the document]
Comment (DeNeutoy) [reply allennlp/tests/training/metrics/spearman_correlation_test.py]: Generally if these tests could be simpler (no striding, just test the case with and without masks for random tensors). In particular, the tests do not call the metric with _batched_ tensors, which would be good (rather than having the doubly nested for loops).

File: allennlp/tests/training/metrics/spearman_correlation_test.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+import torch
+import numpy as np
+from numpy.testing import assert_allclose
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.training.metrics import SpearmanCorrelation
+
+
+def spearman_corrcoef(predictions, labels, mask=None):
```
[the remark this answers is not in the document]
Comment (JackKuo666) [reply allennlp/tests/training/metrics/spearman_correlation_test.py]: ok

File: allennlp/tests/training/metrics/spearman_correlation_test.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        total += (predictions[i][0] - labels[i][0]) ** 2
+    expected_spearman_correlation = 1 - float(6 * total) / (n * (n ** 2 - 1))
+
+    return expected_spearman_correlation
+
+
+class SpearmanCorrelationTest(AllenNlpTestCase):
+    def test_spearman_correlation_unmasked_computation(self):
+        spearman_correlation = SpearmanCorrelation()
+        batch_size = 100
```
[the remark this answers is not in the document]
Comment (JackKuo666) [reply allennlp/tests/training/metrics/spearman_correlation_test.py]: ok

File: allennlp/tests/training/metrics/spearman_correlation_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JackKuo666) [reply allennlp/tests/training/metrics/spearman_correlation_test.py]: ok

File: allennlp/training/metrics/spearman_correlation.py (hunk 1 above)
Comment (JackKuo666) [reply allennlp/training/metrics/spearman_correlation.py]: ok

**Comment by schmmd** [comment]
Thanks @JackKuo666!  Do you think this is good to merge now?

**Comment by JackKuo666** [comment]
Yes, thanks for your review!
